execution_engine_testing/tests/src/test/contract_messages.rs

```diff
@@ -494,3 +542,241 @@ fn should_not_exceed_configured_limits() {
+    assert_eq!(
+        U512::from(written_size_expected * GAS_PER_BYTE_COST as usize),
+        emit_message_gas_cost
+    );
+}
+
+#[ignore]
+#[test]
+fn should_charge_increasing_gas_cost_for_multiple_messages_emitted() {
+    const FIRST_MESSAGE_EMIT_COST: u32 = 100;
+    const COST_INCREASE_PER_MESSAGE: u32 = 50;
+    const MESSAGES_TO_EMIT: u32 = 4;
+    const EMIT_MULTIPLE_EXPECTED_COST: u32 = FIRST_MESSAGE_EMIT_COST * MESSAGES_TO_EMIT
+        + (MESSAGES_TO_EMIT - 1) * MESSAGES_TO_EMIT / 2 * COST_INCREASE_PER_MESSAGE;
+    const EMIT_MESSAGES_FROM_MULTIPLE_CONTRACTS: u32 =
+        FIRST_MESSAGE_EMIT_COST * 3 + 3 * COST_INCREASE_PER_MESSAGE; // simplification of above
+
+    let wasm_config = WasmConfig::new(
+        DEFAULT_WASM_MAX_MEMORY,
+        DEFAULT_MAX_STACK_HEIGHT,
+        OpcodeCosts::zero(),
+        StorageCosts::zero(),
+        HostFunctionCosts::zero(),
+        MessageLimits::default(),
+        MessageCosts {
+            first_message_cost: FIRST_MESSAGE_EMIT_COST,
+            cost_increase_per_message: COST_INCREASE_PER_MESSAGE,
+        },
+    );
+
+    let custom_engine_config = EngineConfigBuilder::default()
+        .with_wasm_config(wasm_config)
+        .build();
+
+    let builder = RefCell::new(LmdbWasmTestBuilder::new_temporary_with_config(
+        custom_engine_config,
+    ));
+    builder
+        .borrow_mut()
+        .run_genesis(&PRODUCTION_RUN_GENESIS_REQUEST);
+
+    let contract_hash = install_messages_emitter_contract(&builder);
+
+    // Emit one message in this execution. Cost should be `FIRST_MESSAGE_EMIT_COST`.
+    emit_message_with_suffix(&builder, "test", &contract_hash, DEFAULT_BLOCK_TIME);
+    let emit_message_gas_cost = builder.borrow().last_exec_gas_cost().value();
+    assert_eq!(emit_message_gas_cost, FIRST_MESSAGE_EMIT_COST.into());
+
+    // Emit multiple messages in this execution. Cost should increase for each message emitted.
+    let emit_messages_request = ExecuteRequestBuilder::contract_call_by_hash(
+        *DEFAULT_ACCOUNT_ADDR,
+        contract_hash,
+        ENTRY_POINT_EMIT_MULTIPLE_MESSAGES,
+        runtime_args! {
+            ARG_NUM_MESSAGES_TO_EMIT => MESSAGES_TO_EMIT,
+        },
+    )
+    .build();
+    builder
+        .borrow_mut()
+        .exec(emit_messages_request)
+        .expect_success()
+        .commit();
+
+    let emit_multiple_messages_cost = builder.borrow().last_exec_gas_cost().value();
+    assert_eq!(
+        emit_multiple_messages_cost,
+        EMIT_MULTIPLE_EXPECTED_COST.into()
+    );
+
+    // Try another execution where we emit a single message.
+    // Cost should be `FIRST_MESSAGE_EMIT_COST`
+    emit_message_with_suffix(&builder, "test", &contract_hash, DEFAULT_BLOCK_TIME);
+    let emit_message_gas_cost = builder.borrow().last_exec_gas_cost().value();
+    assert_eq!(emit_message_gas_cost, FIRST_MESSAGE_EMIT_COST.into());
+
+    // Check gas cost when multiple messages are emitted from different contracts.
```

Comment:
Moving it will add a lot of boiler plate which I don't think it's warranted in this situation. I would prefer to leave the check in this test which is related to increasing gas costs.

Reply:
Ok, fair enough.
Please just confirm if my understanding of this test scenario is correct:
```
Step 1:
1. Emit one message via "emit_message" entry point in the original contract
2. Expect: FIRST_MESSAGE_EMIT_COST

Step 2:
1. Emit 4 messages from single contract via "emit_multiple_messages" entry point in the original contract
2. Expect: emit_cost_per_execution(4)

Step 3:
1. Again emit one message via "emit_message" entry point in the original contract (to assert that costs calculation is reset across executions)
2. Expect: emit_cost_per_execution(4)

Step 4:
1. Emit 3 messages 
    - two via "emit_message_from_each_version" entry point in the upgraded contract
    - one via "emit_message" entry point in the original contract
2. Expect: emit_cost_per_execution(3)
```